lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+  // Get the parents done counter for the node. It should have
+  // been created in the constructor.
+  auto it = nodeParentsDone_.find(node);
+
+  if (it == nodeParentsDone_.end()) {
+    assert(false && "Node parents done counter should exist but not found!");
+  }
+
+  // fetch_add must be used here so that the function returns true to only
+  // one caller.
+  unsigned numParents = (node->parents).size();
+  unsigned previousValue = (it->second).fetch_add(increment);
+  unsigned newValue = previousValue + increment;
+
+  // The new value of the counter cannot exceed the number of parents that
+  // the node has.
+  if (newValue > numParents) {
+    assert(false && "Node parents done counter incremented beyond limit!");
+  }
+
+  // Return true only when the counter hits the total numer of parents.
+  return (newValue == numParents);
+}
+
+void ExecutionState::insertIntoResultCtx(Placeholder *P, Tensor &&T) {
+  assert(resultCtx_ && "Execution result context should exist!");
+  std::lock_guard<std::mutex> lock(contextMtx_);
+  resultCtx_->insert(P, std::move(T));
+}
+
+std::unique_ptr<Context> ExecutionState::getUniqueResultContextPtr() {
+  assert(resultCtx_ && "Execution result context should exist!");
+  return std::move(resultCtx_);
+}
+
+void ExecutionState::setResultCode(const ResultCode resultCode) {
+  if (resultCode_ != ResultCode::Failed) {
+    resultCode_ = resultCode;
+  }
+}
+
+void ThreadPoolExecutor::run(const DAGNode *root,
+                             std::unique_ptr<Context> context,
+                             RunIdentifierTy runId, DoneCbTy cb) {
+  // If list of roots is empty, there is nothing to do. Give back the
+  // context so the caller can reuse it.
+  if (!root) {
+    cb(runId, ResultCode::Executed, std::move(context));
+    return;
+  }
+
+  // If the given run ID corresponds to a run already in progress, there is
+  // also nothing to do, but return an error. Give back the context so the
+  // caller can reuse it.
+  std::unique_lock<std::mutex> lock(executionStatesMutex_);
+  if (executionStates_.find(runId) != executionStates_.end()) {
+    cb(runId, ResultCode::Failed, std::move(context));
+    return;
+  }
+
+  // Create execution state tracker object for this run ID.
+  auto executionState =
+      std::make_shared<ExecutionState>(runId, root, std::move(cb));
+  executionStates_.insert(std::make_pair(runId, executionState));
+  lock.unlock();
+
+  // Execute all child nodes of root.
+
+  // Mark the child nodes as "inflight" (i.e. currently executing). This must be
+  // done here instead of inside executeDAGNode() so that a node can be
+  // executed while placeholders are being propagated for the next node without
+  // the callback for that node deleting the execution state.
+  executionState->incrementInflightNodes((root->children).size());
+
+  // Convert the Context to shared_ptr.
+  std::shared_ptr<Context> sharedContext = std::move(context);
```

Comment:
Looks like we don't store sharedContext anywhere, are you just doing this to pass them into multiple functions? Feels safe to use a raw pointer here.

Reply:
So this is just a downstream effect of having to convert the `unique_ptr<Context>` returned by the `DeviceManager` to a `shared_ptr<Context>` (see line 282) in order to move the handling of the result onto the `ThreadPoolExecutor` thread pool. Apparently you can't capture `unique_ptr` in a lambda?

Reply:
You can, see above (or check out the DeviceManager implementations where we do it).